tests/test_utils.py

```diff
@@ -254,3 +257,60 @@ def test_data_adapter(drop_missing: bool):
     assert (result["C"] != 7).all()
     assert (result["C"] != 9).all()
     assert (result["D"] == 7).all()
+
+
+@pytest.mark.parametrize("reporting_level", [LogLevel.DEBUG, LogLevel.INFO])
+@pytest.mark.parametrize("loglevel", [LogLevel.DEBUG, LogLevel.INFO])
+@pytest.mark.parametrize("tag_func_name", [True, False])
+def test_runtime_decorator(caplog, reporting_level, loglevel, tag_func_name):
+    """
+    Test that run_time_metrics_decorator reports correct information for every run of the algorithm.
+
+    Firstly tests that wrapped method executes even when no logger is passed
+    Secondly tests that wrapped method sends logs when logger is passed.
+
+    :param caplog: pytest fixture for testing logging.
+    :param reporting_level: Reporting level defining at what level decorator sends logs.
+    :param loglevel: Loglevel that is tested.
+    """
+    logger = SemanticLogger().add_log_source(
+        log_source_name="decorator_test",
+        min_log_level=loglevel,
+        log_handlers=[StreamHandler(sys.stdout)],
+        is_default=True,
+    )
+
+    class DummyMetricProvider:
+        def gauge(self, metric_name, metric_value, tags):
+            """Dummy provider to assert passed values"""
+            assert metric_name == run_type
+            assert type(metric_value) == float
+            assert ~tag_func_name or tags["function_name"] == "test_function"
```

Comment:
This is not an error per se, but usually tilde (~) is used for bit string operations and negation (not) is used for Boolean arithmetic like here